Automattic/jetpack
# Adding Gutenberg support to Publicize

This Pull Request contains the changes from #9144 but on a branch local to this repo.


==== Original description follows ===

Fixes #9039 (No Publicize form in Gutenberg)
Dependencies:
- PR #9466 for Jetpack connection data access
- [Gutenberg PR #5795](https://github.com/WordPress/gutenberg/pull/5795) for Gutenberg extensibility


## Changes proposed in this Pull Request:
#### Summary
Bringing Jetpack's Publicize feature to the Gutenberg editor! ---
![image](https://user-images.githubusercontent.com/6209518/39265615-93bd9ba0-487c-11e8-8b3b-c297f53e9089.png)

Without this PR, Publicize invisibly shares all new Gutenberg posts to all connected social accounts.

#### Front-End Design Note
The design is based heavily on @MichaelArestad's [post-publish design concept](https://automattic.invisionapp.com/share/2EGBADFXKAD#/screens/284765844_Start). The desired long-term direction for Publicize is to incorporate it in the post-publish panel (after the user publishes the post), but on this PR, as a first iteration, it is being placed in the pre-publish panel.

#### Architecture Notes
This PR touches a lot of files and there's several early commits that were taking the design in a different direction. In an effort to make review easier, I'll try to distill the list changes into the most important new content and describe how that's hooked into existing content:

  - This PR implements the Publicize form as React-based UI, with components implemented in newly created `modules\publicize\assets\gutenberg_client` directory
  - Publicize form is hooked into Gutenberg via [Slot-Fill](https://github.com/WordPress/gutenberg/tree/master/components/slot-fill#usage)
     - The Publicize `Fill` UI is created on this PR ([in `publicize-gutenberg-plugin.jsx`](https://github.com/Automattic/jetpack/pull/9144/files#diff-3264bcd048332c7a45fcfdb928d6c526)) and the pre-publish `Slot` is created in [Gutenberg PR #5795](https://github.com/WordPress/gutenberg/pull/5795)
   -  Newly created `jetpack\modules\publicize\class-jetpack-publicize-gutenberg.php` does the setup for the new UI.
       - Registers new UI JS to be loaded with Gutenberg editor
       - Registers REST field on the `post` endpoint to pass in Publicize `title` and `connections` form options when saving post:
![image](https://user-images.githubusercontent.com/6209518/38937305-c4e4a656-42d7-11e8-8996-12632ab3b15f.png) 
       - Registers new Ajax endpoint `wp_ajax_get_publicize_connections` to expose connection list so client can refresh connections
   - Depends on `try/publicize-decouple-ui` (PR #9466) to expose connection data access methods

## Testing instructions:



1. Checkout this branch and merge dependended on branch:
    `git checkout add/gutenberg-publicize && git merge try/publicize-decouple-ui --no-commit`
2. Checkout [Gutenberg PR #5795](https://github.com/WordPress/gutenberg/pull/5795) and build to enable pre-publish extensibility
3. Activate Gutenberg and Jetpack
4. Create new post
5. Click "Publish" to get Publicize form in pre-publish sidebar
6. Try stuff:
    - Add connections
    - Open settings
    - Edit message
    - Check/uncheck connections
    - Refresh connections
    - Publish post
    - Check social accounts for post and message
    - Try changing previously published/shared post to draft and re-sharing (Publicize should be disabled)
    - Try Publicize in classic editor to make sure nothing changed
